### ISSUE TITLE
Constructor / method in hover

### ISSUE THREAD

**Issue by guibou** [haskell/haskell-language-server#3326]
## Is your enhancement request related to a problem? Please describe.

Regularly, I'm using the hover command to look at the type of an identifier, class, ...

For function, I do get a lot of useful information, such as the type of the function, the resolved type, as well as precedence (if defined), for example, with the following code:

```haskell
data Test a = A | B a | C

fiz :: Test a -> Test a
fiz x = x

infixr 7 `fiz` 

g = fiz (B True)
```

I get this informations in hover on `fiz` call:

```
fiz :: forall a. Test a -> Test a
───────────────────────────────────────
*Defined at /home/guillaume/Foo.hs:5:1*

* * *

_ :: Test Bool -> Test Bool
───────────────────────────────────────
* * *  
infixr 7 `fiz`  
```

However, for type and class, I have not much information, except the documentation and kind, e.g:

For hover on the type `Test`:

```
Test :: * -> *
───────────────────────────────────────
*Defined at /home/guillaume/Foo.hs:2:1*
```

For hover on the constructor `B`:

```
B :: forall a. a %1 -> Test a
────────────────────────────────────────
*Defined at /home/guillaume/Foo.hs:2:19*

* * *

_ :: Bool -> Test Bool
```

And for hover on the `Functor`  class:

```
Functor :: (* -> *) -> Constraint
────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
*Defined in ‘GHC.Base’* *(base-4.15.1.0)*



A type  `f`  is a Functor if it provides a function  `fmap`  which, given any types  `a`  and  `b` 
lets you apply any function from  `(a -> b)`  to turn an  `f a`  into an  `f b` , preserving the
structure of  `f` . Furthermore  `f`  needs to adhere to the following: 
+ **Identity**: `fmap   id  ==  id`
+ **Composition**: `fmap  (f . g) ==  fmap  f .  fmap  g`
 

Note, that the second law follows from the free theorem of the type  `fmap`  and
the first law, so you need only check that the former condition holds.

[Documentation](file:///nix/store/5qyjmd1c52bzhbyxcy3kll5hi4zlwbnm-ghc-9.0.2-doc/share/doc/ghc/html/libraries/base-4.15.1.0/GHC-Base.html#t:Functor)

[Source](file:///nix/store/5qyjmd1c52bzhbyxcy3kll5hi4zlwbnm-ghc-9.0.2-doc/share/doc/ghc/html/libraries/base-4.15.1.0/src/GHC.Base.html#Functor)
```

I would like more information.

## Describe the solution you'd like

For type: having the listing of the different constructors.
For constructor: the listing of the other constructors may be interesting.
For class: the listing of the different method,s with the minimal required implementation.

## Describe alternatives you've considered

Right now my alternative is to open the documentation manually and look for the informations I care about.

**Comment by michaelpj** [comment]
So, the general heuristic is that we give information relating to the _declaration_ of an entity in the hover. i.e. 
- for functions you get the type signature+docs, but not the definition
- for datatypes you get the kind signature+docs, but not the definition
- for classes you get the kind signature+docs, but not the definition

So a way of understanding what you're asking for is that you _want_ us to include some of the definition in hovers, in some cases.

That reminded me of this issue: https://github.com/haskell/haskell-language-server/issues/2222

I _think_ I still agree with that: this is a generic problem ("I want to see some of the definition of the thing at point"), and many clients provide functionality to do just that ("peek at definition"). So arguably this is something that can (should?) be done entirely client-side.

**Comment by guibou** [comment]
I agree with you. I actually found a bit of informations related "peek definition" for vim here: https://github.com/neovim/neovim/pull/12368 (Adding for reference). I'm still looking for a nice plugin.

There are still one issue with this approach, is that the `peek` feature relies on the server to provide the correct location information. This seem to work fine with HLS for files which are part of the project, but not otherwise, which can be understood, considering that the source are not available.

Do you think it may be possible to extract the haddock informations and point to the source in haddock (if they exist)?

**Comment by michaelpj** [comment]
Well, we also wouldn't be able to show the RHS in the hover in that case either!

> Do you think it may be possible to extract the haddock informations and point to the source in haddock (if they exist)?

I think we already give all the location info we have (it says "defined in <module> from <package>", right?) but maybe there's more we could put in :thinking:
